### PR TITLE
Autocomplete Transaction Description

### LINKED ISSUE #524: Autocomplete transaction description
It would be helpful to have an auto-complete feature for transaction descriptions based on existing transactions when you start typing. These suggestions would be based on previous transactions with similar descriptions.

Additionally, it would be nice if these transactions could also display the group to which previous transactions with these descriptions belong. This group can be assigned when selecting a suggestion.

### LINKED ISSUE #551: ID widget doesn't restore going from small window to maximize
# Small To Maximize

[Screencast from 2023-06-06 02-02-21.webm](https://github.com/NickvisionApps/Denaro/assets/17648453/58c4dfce-1f93-4403-9e45-9dcfeab96919)

### PR DESCRIPTION
**Pull request by nlogozzo**
Closes #524 
Closes #551

### CODE REVIEW

**Comment by nlogozzo** [comment]
[Screencast from 2023-06-06 01-51-56.webm](https://github.com/NickvisionApps/Denaro/assets/17648453/3f25be8e-18e0-4564-9166-18bee15f6f6d)

@fsobolev Ok so everything is implemented for auto-completion of transaction description and as you can see from the video, it works and works well. The only issue is the rows expand the height of the dialog and at times can get too large. 
Do you have any other design ideas to implement this/fix this? Like I said all the code is there so feel free to play around with the design to get it to work :)

**Comment by nlogozzo** [comment]
Fixed #551 with https://github.com/NickvisionApps/Denaro/pull/550/commits/83ce95497db727d794eac5df4f4a0bcc6e859d65

File: NickvisionMoney.GNOME/Views/TransactionDialog.cs [hunk 1]
```diff
@@ -480,6 +469,34 @@ private void Validate()
...
+            }
+            else
+            {
+                _autocompleteDialog.Hide();
+            }
+        }
+        else
+        {
+            _autocompleteDialog.Hide();
+        }
```
Comment (DaPigGuy) [inline NickvisionMoney.GNOME/Views/TransactionDialog.cs:497]:
```suggestion
                return;
            }
        }
         _autocompleteDialog.Hide();
```

**Comment by nlogozzo** [comment]
Problem I'm having right now, the `AutocompleteDialog` keeps stealing focus away from the `TransactionDialog` and after I call `_autocompleteDialog.Present()` I try adding the following:
```csharp
GrabFocus();
_descriptionRow.GrabFocus();
_descriptionRow.SetPosition(-1);
```
but it doesn't seem to help...

**Comment by nlogozzo** [comment]
Might have found a different API to use...

**Comment by nlogozzo** [comment]
> Might have found a different API to use...

Was gonna use GtkTooltip but never mind, doesn't work.

**Comment by nlogozzo** [comment]
Got it working 🎉🎉:
[Screencast from 2023-06-07 01-26-39.webm](https://github.com/NickvisionApps/Denaro/assets/17648453/09711249-8a86-456e-8f7e-4c794386425e)

Also limited the number of suggestions to the first 5 results as to not overcrowd the overlay

**Comment by nlogozzo** [comment]
> Additionally, it would be nice if these transactions could also display the group to which previous transactions with these descriptions belong. This group can be assigned when selecting a suggestion.

Going to implement this part of the feature request now as well....

**Comment by nlogozzo** [comment]
A suggestion will also show the group and inherit the group and color setting if available as well

[Screencast from 2023-06-07 02-40-57.webm](https://github.com/NickvisionApps/Denaro/assets/17648453/97fd3d0d-ce3a-4cbd-b28c-bed99ecf95e3)

**Comment by nlogozzo** [comment]
Now it's ready :)

**Comment by nlogozzo** [comment]
image

**Review by fsobolev — APPROVED** [review]

**Pull request merged**